File: utility/thermal_simulation/control/pid_control.py
class PidControl:
    def __init__(self, 
        Kp, 
        Ki, 
        Kd, 
        min_output=0.0, 
        max_output=1.0
    ):
        self.Kp = Kp
        self.Ki = Ki
        self.Kd = Kd

        self.min_output = min_output
        self.max_output = max_output

        self.integral = 0.0
        self.prev_error = 0.0
        self.output = 0.0

        self._first = True

        
    def reset(self):
        self.integral = 0
        self._first = True

    
    def step(self, measured, target, dt):

        error = target - measured

        # Integral
        old_integral = self.integral
        self.integral += error * dt

        # Derivative (simple backward difference)
        if self._first:
            derivative = 0.0
            self._first = False
        else:
            derivative = (error - self.prev_error) / dt

        self.prev_error = error

        # Raw PID output
        u = (
            self.Kp * error +
            self.Ki * self.integral +
            self.Kd * derivative
        )

        # saturate output
        u = min(self.max_output, max(self.min_output, u))

        # prevent further windup
        if(u == self.min_output or u == self.max_output):
            if error > 0:
                self.integral = old_integral
            if error < 0:
                self.integral = old_integral

        self.output = u
        return self.output

File: src/core/util/control/PidControl.cpp
#include "PidControl.h"

namespace Garbox {

PidControl::PidControl(
    float Kp,
    float Ki,
    float Kd,
    float minOutput,
    float maxOutput
):
    // initialize members
    mKp(Kp),
    mKi(Ki),
    mKd(Kd),
    mMinOutput(minOutput),
    mMaxOutput(maxOutput)
{
    // constructor body
}

void PidControl::reset(){
    mOutput = 0.0f;
    mIntegral = 0.0f;
    mPrevError = 0.0f;
    mFirst = true;
}

float PidControl::step(float measured, float target, float dtSeconds){

    // compute error
    const float error = target - measured;

    // store old integral for anti windup
    const float oldIntegral = mIntegral;

    // integrate
    mIntegral += error * dtSeconds;

    // derivative
    float derivative = 0.0f;
    if(mFirst){
        derivative = 0.0f;
        mFirst = false;
    }
    else {
        derivative = (error - mPrevError) / dtSeconds;
    }

    mPrevError = error;

    // compute raw output
    float u = (mKp * error + mKi * mIntegral + mKd * derivative);

    // saturate output
    if(u > mMaxOutput){
        u = mMaxOutput;
    }
    else if(u < mMinOutput){
        u = mMinOutput;
    }

    // anti windup
    const bool saturated = (u == mMinOutput) || (u == mMaxOutput);
    if(saturated){
        if(error > 0.0f){
            mIntegral = oldIntegral;
        }
        else if(error < 0.0f){
            mIntegral = oldIntegral;
        }
    }

    mOutput = u;
    return mOutput;
}

} // namespace Garbox
